File: src/BomberMan/brain_logic.py
import random
import brain
b = brain.IA()

actions = [0, 1, 2, 3, 4, 5]

def heuristic(id):
    return ((b.getNumberOfBoxesFromId(0) - b.getNumberOfBoxesFromId(id)) * 10)
    + ((b.getNumberOfPowerUpFromId(id) - b.getNumberOfPowerUpFromId(0)) * 100)
    + ((b.getNumberOfPlayerFromId(0) - b.getNumberOfPlayerFromId(id)) * 1000)

def calculate(id, depth):
    if depth == b.getMaxDepth():
        b.doUpdateFromId(id, 200)
        if b.amIDeadFromId(id):
            return -1
        return heuristic(id)

    res = []
    for action in actions:
        if b.amIDeadFromId(id):
            res += [-1]
        else:
            if b.isActionPossible(id, action):
                r = b.copyFromRefereeId(id)
                b.doActionFromId(r, action)
                b.doUpdateFromId(r, 0)
                h = calculate(r, depth + 1)
                res += [h + heuristic(id)]
            else:
                res += [-1]
    m = max(res)
    if depth == 0:
        ret = [i for i, j in enumerate(res) if j == m]
        if len(ret) == 1:
            return ret[0]
        if m == 0:
            ret = filter(lambda a: a != 0, ret)
        return random.choice(ret)
    return m

def execute():
    return calculate(0, 0)

File: src/BomberMan/brain_logic.lua
function heuristic(me, id)
    return ((getNumberOfBoxes(me, 0) - getNumberOfBoxes(me, id)) * 10) + ((getNumberOfPowerUp(me, id) - getNumberOfPowerUp(me, 0)) * 100) + ((getNumberOfPlayer(me, 0) - getNumberOfPlayer(me, id)) * 1000)
end

function calculate(me, id, depth)
    if depth == getMaxDepth(me) then
        doUpdate(me, id, 200)
        if amIDead(me, id) then
            return -1
        end
        return heuristic(me, id)
    end

    local res = {}
    for action=0,5 do
        if amIDead(me, id) then
            res[#res+1] = -1
        else
            if isActionPossible(me, id, action) then
                local r = copy(me, id)
                doAction(me, r, action)
                doUpdate(me, r, 0)
                res[#res+1] = calculate(me, r, depth + 1) + heuristic(me, id)
            else
                res[#res+1] = -1
            end
        end
    end

    local m = math.max(table.unpack(res))
    if depth == 0 then
        local ret = {}
        for k in pairs (res) do
            if res[k] == m then
                ret[#ret+1] = k - 1
            end
        end
        if #ret == 1 then
            return ret[1]
        end
        if m == 0 then
            for k in pairs (ret) do
                if ret[k] == 0 then
                    ret[k] = nil
                end
            end
        end
        return ret[math.random(#ret)]
    end
    return m
end

brain = function(me)
    return calculate(me, 0, 0)
end